cli/src/commands/sign.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write;
+
+use jj_lib::object_id::ObjectId;
+use jj_lib::signing::SignBehavior;
+
+use crate::cli_util::{user_error, CommandError, CommandHelper, RevisionArg};
+use crate::ui::Ui;
+
+/// Cryptographically sign a revision
+#[derive(clap::Args, Clone, Debug)]
+pub struct SignArgs {
+    /// What key to use, depends on the configured signing backend.
+    #[arg()]
+    key: Option<String>,
+    /// What revision to sign
+    #[arg(long, short, default_value = "@")]
+    revision: RevisionArg,
+    /// Sign a commit that is not authored by you or was already signed.
+    #[arg(long, short)]
+    force: bool,
+    /// Drop the signature, explicitly "un-signing" the commit.
+    #[arg(long, short = 'D', conflicts_with = "force")]
+    drop: bool,
+}
+
+pub fn cmd_sign(ui: &mut Ui, command: &CommandHelper, args: &SignArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    let commit = workspace_command.resolve_single_rev(&args.revision)?;
```

Comment:
would be nice to accept multiple revs

Reply:
Descendants of rewritten commits (and signing is rewriting) get rebased, so if one of the revisions is a descendant of another there are issues you'd need to figure out how to avoid.

Another thing is a rebase optimisation that's needed for instances where you wan't to sign a range of commits - currently (if you do it in a loop in a shell, I guess) if you sign 10 commits you'd have 45 (ish? should be a triangle, n(n-1)/2) useless commits created from all the rebases.

So initially we decided to at least have a single-revision sign.

Also technically any rewriting command with `--sign` global flag is a sign.  
Hm, hah, a noop rebase with --sign should be the rebase optimisation I was talking about above, interesting

Reply:
I hope to make it easier to do such transformations while rebasing. It's just been hard to find time for it. The idea is that we'll something similar to `rebase_descendants()` but instead `transform_descendants(callback)` where the callback gets a `CommitBuilder` populated with the default values (so `rebase_descendants` will be `transform_descendants(|_|) {}`).